modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceInfo.java

```diff
@@ -116,15 +125,28 @@ public IgniteUuid serviceId() {
     }
 
     /** {@inheritDoc} */
-    @SuppressWarnings("unchecked")
     @Override public Class<? extends Service> serviceClass() {
         if (cfg instanceof LazyServiceConfiguration) {
+            if (srvcCls != null)
+                return srvcCls;
+
             String clsName = ((LazyServiceConfiguration)cfg).serviceClassName();
 
             try {
-                return (Class<? extends Service>)Class.forName(clsName);
+                srvcCls = (Class<? extends Service>)Class.forName(clsName);
+
+                return srvcCls;
             }
             catch (ClassNotFoundException e) {
+                GridDeployment srvcDep = ctx.deploy().getDeployment(clsName);
```

Comment:
We have similar logic here `IgniteServiceProcessor#copyAndInject`

Can't we reuse it?

Reply:
It's only one line of code can be reused, I think it's not such a good idea to make a new method to reuse only `ctx.deploy().getDeployment(clsName)`.

Reply:
I didn't mean copy/paste :)

I thought that there is another way to load className.

For example, load and set a className externally (without KernalContext field in `ServiceInfo`).
Not sure if `IgniteServiceProcessor#copyAndInject` is a suitable place for that or not.

Reply:
I've tried to implement class resolving logic in `IgniteServiceProcessor` before `ServiceInfo` constructor and pass it as an argument, but I think it's risky if we fail in `processServicesChangeRequest`, so I've moved class resolving logic to  `ServiceInfo.serviceClass()` method.